tests/qubit_codes/test_graphstates.py

```diff
@@ -71,8 +94,115 @@ def test_index(self, random_graph):
         E_adj = E.adj_mat
         assert np.array_equal(H_adj, E_adj)
 
-    # def test_macronode(self):
-    # pass
+    def test_add_qubit(self, random_graph_3D):
+        """Test the add_qubit function on a random EGraph."""
+
+        E = EGraph(random_graph_3D)
+        E.adj_generator()
+        max_ind_old = max(E.to_points)
+        n_old = E.number_of_nodes()
+        E.add_qubit()
+        E.add_qubit()
+        assert n_old + 2 == E.number_of_nodes()
+        assert max(E.to_points) == max_ind_old + 2
+        assert E.adj_mat is None
+
+        E = EGraph(random_graph_3D)
+        E.index_generator()
+        E.add_qubit(qubit=(100, 101, -400))
+        assert (100, 101, -400) in E
+
+        with pytest.raises(Exception) as e:
+            E.add_qubit(qubit=(1, 1, 1, 1))
+        assert e.type == ValueError
+
+        with pytest.raises(Exception) as e:
+            E.add_qubit(qubit=0)
+        assert e.type == TypeError
+
+        with pytest.raises(Exception) as e:
+            E.add_qubit(qubit=(1, 1, 100), neighbors=["a", "b"])
+        assert e.type == TypeError
+
+        E = EGraph(random_graph_3D)
+        n_edges_old = E.number_of_edges()
+        E.index_generator()
+        E.add_qubit(neighbors=[0, 1])
+        assert n_edges_old + 2 == E.number_of_edges()
+
+        E = EGraph(random_graph_3D)
+        n_edges_old = E.number_of_edges()
+        new_neighs = [neigh for neigh in E.nodes() if rng.integers(0, 10) > 7]
+        E.add_qubit(neighbors=new_neighs)
+        assert n_edges_old + len(new_neighs) == E.number_of_edges()
+
+    def test_remove_qubit(self, random_graph_3D):
+        """Test the remove_qubit function on a random EGraph."""
 
-    # def test_slice_coords(self):
-    # pass
+        E = EGraph(random_graph_3D)
+        E.index_generator()
+        to_points_old = E.to_points.copy()
+        n_old = E.number_of_nodes()
+        E.remove_qubit(1)
+        assert n_old == E.number_of_nodes() + 1
+        assert E.adj_mat is None
+        assert to_points_old[2] == E.to_points[2]
+
+        E = EGraph(random_graph_3D)
+        n_old = E.number_of_nodes()
+        nodes = list(E.nodes())
+        E.remove_qubit(nodes[0])
+        E.remove_qubit(nodes[1])
+        assert n_old == E.number_of_nodes() + 2
+        assert E.adj_mat is None
+
+        E = EGraph(random_graph_3D)
+        with pytest.raises(Exception) as e:
+            E.remove_qubit("s")
+        assert e.type == ValueError
+
+        E = EGraph(random_graph_3D)
+        E.index_generator()
+        with pytest.raises(Exception) as e:
+            E.remove_qubit("s")
+        assert e.type == TypeError
+
+    def test_add_qubit_macro(self, random_graph_3D):
```

Comment:
This and test below will change a little with the new add_to_macro option

Reply:
updated and working ✅